common/pooll2tx.go

```diff
@@ -410,3 +422,165 @@ const (
+		toIdx := idxToHez(tx.ToIdx, tx.TokenSymbol)
+		toMarshal.ToIdx = &toIdx
+	}
+	if tx.ToEthAddr != EmptyAddr {
+		toEth := ethAddrToHez(tx.ToEthAddr)
+		toMarshal.ToEthAddr = &toEth
+	}
+	if tx.ToBJJ != EmptyBJJComp {
+		toBJJ := bjjToString(tx.ToBJJ)
+		toMarshal.ToBJJ = &toBJJ
+	}
+	// Set Rq fields
+	if tx.RqFromIdx != 0 {
+		if tx.RqTokenSymbol == "" {
+			return nil, errors.New("Invalid tx.RqTokenSymbol")
+		}
+		rqFromIdx := idxToHez(tx.RqFromIdx, tx.TokenSymbol)
+		toMarshal.RqFromIdx = &rqFromIdx
+		toMarshal.RqTokenID = &tx.RqTokenID
+		rqAmount := tx.RqAmount.String()
+		toMarshal.RqAmount = &rqAmount
+		toMarshal.RqNonce = &tx.RqNonce
+		toMarshal.RqFee = &tx.RqFee
+		if tx.RqToIdx != 0 {
+			rqToIdx := idxToHez(tx.RqToIdx, tx.RqTokenSymbol)
+			toMarshal.RqToIdx = &rqToIdx
+		}
+		if tx.RqToEthAddr != EmptyAddr {
+			rqToEth := ethAddrToHez(tx.RqToEthAddr)
+			toMarshal.RqToEthAddr = &rqToEth
+		}
+		if tx.RqToBJJ != EmptyBJJComp {
+			rqToBJJ := bjjToString(tx.RqToBJJ)
+			toMarshal.RqToBJJ = &rqToBJJ
+		}
+	}
+	return json.Marshal(toMarshal)
+}
+
+// UnmarshalJSON unmarshals a PoolL2Tx that has been formated in json as specified in `api/swagger.yml:schemas/PostPoolL2Transaction`.
+// Note that ClientIP is not included as part of the json and will be set to "".
+// If State is not setted (State == ""), it will be set to PoolL2TxStatePending.
+func (tx *PoolL2Tx) UnmarshalJSON(data []byte) error {
+	receivedJSON := struct {
+		TxID        TxID                  `json:"id" binding:"required"`
+		Type        TxType                `json:"type" binding:"required"`
+		TokenID     TokenID               `json:"tokenId"`
+		FromIdx     StrHezIdx             `json:"fromAccountIndex" binding:"required"`
+		ToIdx       StrHezIdx             `json:"toAccountIndex"`
+		ToEthAddr   StrHezEthAddr         `json:"toHezEthereumAddress"`
+		ToBJJ       StrHezBJJ             `json:"toBjj"`
+		Amount      *StrBigInt            `json:"amount" binding:"required"`
+		Fee         FeeSelector           `json:"fee"`
+		Nonce       Nonce                 `json:"nonce"`
+		Signature   babyjub.SignatureComp `json:"signature" binding:"required"`
+		RqFromIdx   StrHezIdx             `json:"requestFromAccountIndex"`
+		RqToIdx     StrHezIdx             `json:"requestToAccountIndex"`
+		RqToEthAddr StrHezEthAddr         `json:"requestToHezEthereumAddress"`
+		RqToBJJ     StrHezBJJ             `json:"requestToBjj"`
+		RqTokenID   TokenID               `json:"requestTokenId"`
+		RqAmount    *StrBigInt            `json:"requestAmount"`
+		RqFee       FeeSelector           `json:"requestFee"`
+		RqNonce     Nonce                 `json:"requestNonce"`
+		State       string                `json:"state"`
+	}{}
+	if err := json.Unmarshal(data, &receivedJSON); err != nil {
+		return err
+	}
+	// Check state
+	state := PoolL2TxStatePending
+	if receivedJSON.State != "" {
+		state = PoolL2TxState(receivedJSON.State)
+		if !state.IsValid() {
+			return errors.New("Invalid state")
```

Comment:
Maybe this way log message be better?
```suggestion
			return fmt.Errorf("Invalid state: %s", state)
```